lib/experimental/class-wp-theme-json-gutenberg.php

```diff
@@ -29,3 +29,56 @@ class WP_Theme_JSON_Gutenberg extends WP_Theme_JSON_6_2 {
 	 */
 	protected static $blocks_metadata = null;
 }
+
+/**
+ * Sanitizes global styles user content removing unsafe rules.
+ *
+ * @since 5.9.0
+ *
+ * @param string $data Post content to filter.
+ * @return string Filtered post content with unsafe rules removed.
+ */
+function gutenberg_filter_global_styles_post( $data ) {
+	$decoded_data        = json_decode( wp_unslash( $data ), true );
+	$json_decoding_error = json_last_error();
+	if (
+		JSON_ERROR_NONE === $json_decoding_error &&
+		is_array( $decoded_data ) &&
+		isset( $decoded_data['isGlobalStylesUserThemeJSON'] ) &&
+		$decoded_data['isGlobalStylesUserThemeJSON']
+	) {
+		unset( $decoded_data['isGlobalStylesUserThemeJSON'] );
+
+		$data_to_encode = WP_Theme_JSON_Gutenberg::remove_insecure_properties( $decoded_data );
+
+		$data_to_encode['isGlobalStylesUserThemeJSON'] = true;
+		return wp_slash( wp_json_encode( $data_to_encode ) );
+	}
+	return $data;
+}
+
+/**
+ * Override core's kses_init_filters hooks for global styles,
+ * and use Gutenberg's version instead. This ensures that
+ * Gutenberg's `remove_insecure_properties` function can be called.
+ *
+ * The hooks are only set if they are already added, which ensures
+ * that global styles is only filtered for users without the `unfiltered_html`
+ * capability.
+ *
+ * This function should not be backported to core.
+ */
+function gutenberg_override_core_kses_init_filters() {
+	if ( has_filter( 'content_save_pre', 'wp_filter_global_styles_post' ) ) {
+		remove_filter( 'content_save_pre', 'wp_filter_global_styles_post', 9 );
+		add_filter( 'content_save_pre', 'gutenberg_filter_global_styles_post', 9 );
+	}
+
+	if ( has_filter( 'content_filtered_save_pre', 'wp_filter_global_styles_post' ) ) {
+		remove_filter( 'content_filtered_save_pre', 'wp_filter_global_styles_post', 9 );
+		add_filter( 'content_filtered_save_pre', 'gutenberg_filter_global_styles_post', 9 );
+	}
```

Comment:
Nitpick (non-blocking): `remove_filter` returns true if the given callback existed before it was removed, so this can be slightly simplified.
```suggestion
	if ( remove_filter( 'content_save_pre', 'wp_filter_global_styles_post', 9 ) ) {
		add_filter( 'content_save_pre', 'gutenberg_filter_global_styles_post', 9 );
	}

	if ( remove_filter( 'content_filtered_save_pre', 'wp_filter_global_styles_post', 9 ) ) {
		add_filter( 'content_filtered_save_pre', 'gutenberg_filter_global_styles_post', 9 );
	}
```

Although using `has_filter` explicitly might be clearer and avoid potential confusion, so feel free to keep this as-is!

Reply:
Another good question — in this case, I personally prefer the explicit `has_filter` check so that it's clearer at a glance that the Gutenberg filter is only added if a corresponding core one existed. This deviates a little from how we typically register Gutenberg filters in that we usually expect them to always be running, but in this case, it should only be registered for the use case of no `unfiltered_html` capability.

Again, happy to use the simplified approach if folks prefer! 🙂